### PR TITLE
Rolling restart query system

### PR DESCRIPTION
**Pull request by apsubra**

### CODE REVIEW

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java [hunk 1]
```diff
@@ -284,6 +284,32 @@ private void fastHandle(ChannelHandlerContext ctx, FullHttpRequest request, Stri
...
+                } else {
+                    writer.write("[0]");
+                }
+                break;
+            }
+            case "/mqmaster/healthcheck": {
+                if(meshQueryMaster.getIsActive().get()) {
+                    writer.write("[1]");
+                } else {
+                    writer.write("[0]");
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: I'm curious as to why the brackets around the numbers in the output here?

File: hydra-main/src/main/java/com/addthis/hydra/query/MeshQueryMaster.java [hunk 2]
```diff
@@ -126,6 +130,18 @@ public MeshQueryMaster(QueryTracker tracker) throws Exception {
...
+    public AtomicBoolean getIsActive() {
+        return isActive;
+    }
+
+    public boolean deactivateMqMaster(boolean deactivate) {
+        if(deactivate) {
+            // Set mqMaster's isActive field to false
+            return this.isActive.compareAndSet(true, !deactivate);
+        }
+        return this.isActive.compareAndSet(false, true);
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/query/MeshQueryMaster.java]: This whole method could be replaced with `this.isActive.compareAndSet(deactivate, !deactivate)`. What are we doing on the client side if compare and set is not successful?

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java [hunk 3]
```diff
@@ -284,6 +284,32 @@ private void fastHandle(ChannelHandlerContext ctx, FullHttpRequest request, Stri
...
+                        break;
+                    default:  success = false;
+                }
+                if(success) {
+                    writer.write("[1]");
+                } else {
+                    writer.write("[0]");
+                }
+                break;
+            }
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: We should just set deactivate to true or false, and not worry about sending back success or state. If we get an http 200, assume all is well.

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java (hunk 1 above)
Comment (apsubra) [reply hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: The plan was to have the mq masters under a load balancer. The lb we received (at the time, Ops created it for us) was configured to use "[1] or [0]" as the healthcheck.

Comment (apsubra) [reply hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: Edited healthcheck to be Http 200/400 for now, I made a note to look at this once again when there's more clarity on what the lb will look like

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java (hunk 3 above)
Comment (apsubra) [reply hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: Edited deactivate endpoint to just set() and return Http ok/400

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java [hunk 4]
```diff
@@ -284,6 +284,35 @@ private void fastHandle(ChannelHandlerContext ctx, FullHttpRequest request, Stri
                 writer.write(q.getPaths()[0]);
                 break;
             }
+            case "/mqmaster/deactivate": {
+                log.info("Received MeshQueryMaster deactivate request");
+                String shouldDeactivate = kv.getValue("deactivate");
+                switch (shouldDeactivate) {
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: Could throw NPE here.

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java [hunk 5]
```diff
@@ -284,6 +284,35 @@ private void fastHandle(ChannelHandlerContext ctx, FullHttpRequest request, Stri
...
+                                if (ctx.channel().isActive()) {
+                                    sendError(ctx, HttpResponseStatus.BAD_REQUEST);
+                                }
+                                return;
+                }
+                break;
+            }
+            case "/mqmaster/healthcheck": {
+                // Send 400 if mqMaster is not active
+                // Send 200 if mqMaster is active (NO-OP here as response OK is sent below)
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: Good comment!

File: hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java (hunk 4 above)
Comment (apsubra) [reply hydra-main/src/main/java/com/addthis/hydra/query/web/HttpQueryHandler.java]: Added a null-check

**Comment by tedpearson** [comment]
Looks good

**Pull request merged**